### PR TITLE
Loosen CSV download limits

### PR DESCRIPTION
**Pull request by agilliland**
update QP max results limitations in a couple ways:
- set the absolute limit on QP results up to just over 1m rows, which is the max Excel allows.
- allow query execution callers to specify their own :max-results and :max-results-bare-rows values so that we can contextually constrain the result sizes.  this allows us to easily have the api allow 10k results, a pulse to allow 500 results, and a CSV download to allow 1m results.

also included in this will be a modal on the frontend warning users if they are downloading data sets which are larger than the normal limits we enforce through the api.

### CODE REVIEW

**Comment by agilliland** [comment]
@camsaul check this out when you get a minute.

i'm allowing some parts of the QP results limitations to be set dynamically by the query caller so that we can enforce different limits in different situations.  we still enforce an absolute max though, which is bumped up to 1m rows.

the interesting bit is that i'm having the query callers set their limit definitions by adding a `:constraints {}` attribute to the top level of the query before it's passed into the `(process)` function.  this seemed like the most sensible way to pass in information on how the query should be processed, even though these aren't specifically parts of the query.

File: src/metabase/driver/query_processor.clj [hunk 1]
```diff
@@ -70,14 +78,20 @@
...
-  (fn [query]
-    (let [results     (qp query)
-          num-results (count (:rows results))]
+  (fn [{{:keys [max-results max-results-bare-rows]} :constraints, :as query}]
+    (let [results-limit (if (rows-query-without-limits? query)
+                          (or max-results-bare-rows
+                              max-results
+                              absolute-max-results)
+                          (or max-results
+                              absolute-max-results))
```
Comment (camsaul) [inline src/metabase/driver/query_processor.clj]: Might be a little more readable if you do 
```clojure
(or (when (rows-query-without-limits? query)
      max-results-bare-rows)
    max-results
    absolute-max-results)
```

File: test/metabase/driver/query_processor_test.clj [hunk 2]
```diff
@@ -481,12 +488,28 @@
...
-(expect max-result-rows
+;; Apply limit-max-result-rows to an infinite sequence and make sure it gets capped at `absolute-max-results`
+(expect absolute-max-results
   (->> (((u/runtime-resolved-fn 'metabase.driver.query-processor 'limit) identity) {:rows (repeat [:ok])})
        :rows
        count))
 
+;; Apply an arbitrary max-results on the query and ensure our results size is appropriately constrained
+(expect 1234
+  (->> (((u/runtime-resolved-fn 'metabase.driver.query-processor 'limit) identity) {:constraints {:max-results 1234}
```
Comment (camsaul) [inline test/metabase/driver/query_processor_test.clj]: there's not real benefit to using `u/runtime-resolved-fn` here. That's just a way to resolve a fn for repeated use without the overhead of calling `resolve` every time. Just do 

```clojure
@(resolve 'metabase.driver.query-processor 'limit)
```

instead

**Comment by camsaul** [comment]
Works for me but I don't think there's any real benefit to nesting the keys inside `constraints`, especially since there're only two of them. There's no way those are conflicting with anything in the top-level of the query dict so it just makes things a little more verbose when we need to get their values or write queries by hand in the REPL

**Pull request merged**